File: Reactivos/src/reactivos/Escenario.java

package reactivos;

import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ItemEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import javax.swing.BorderFactory;
import javax.swing.ButtonGroup;
import javax.swing.ImageIcon;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.JRadioButtonMenuItem;
import java.io.File;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;

/**
 *
 * @author macario
 */
public class Escenario extends JFrame
{
    
    private JLabel[][] tablero;     
    private int[][] matrix;
    private final int dim = 15;

    private ImageIcon robot1;
    private ImageIcon robot2;
    private ImageIcon obstacleIcon;
    private ImageIcon sampleIcon;
    private ImageIcon actualIcon;
    private ImageIcon motherIcon;
    
    private Agente wallE;
    private Agente eva;
   
    private final BackGroundPanel fondo = new BackGroundPanel(new ImageIcon("Reactivos\\imagenes\\surface.jpg"));
    
    private final JMenu settings = new JMenu("Settigs");
    private final JRadioButtonMenuItem obstacle = new JRadioButtonMenuItem("Obstacle");
    private final JRadioButtonMenuItem sample = new JRadioButtonMenuItem("Sample");
    private final JRadioButtonMenuItem motherShip = new JRadioButtonMenuItem("MotherShip");
    
    public Escenario()
    {
        this.setContentPane(fondo);
        this.setTitle("Agentes");
        this.setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
        this.setBounds(50,50,dim*50+35,dim*50+85);
        initComponents();
        
                // Clip de audio
        try
        {    
            AudioInputStream inputStream = AudioSystem.getAudioInputStream(new File("audio/audio.aiff"));
            Clip clip = AudioSystem.getClip();
            clip.open(inputStream);
            clip.loop(Clip.LOOP_CONTINUOUSLY);
        }
        catch (Exception ex)
        {
            ex.printStackTrace(System.out);
        }
    }
        
    private void initComponents()
    {
        ButtonGroup settingsOptions = new ButtonGroup();
        settingsOptions.add(sample);
        settingsOptions.add(obstacle);       
        settingsOptions.add(motherShip);
        
        JMenuBar barraMenus = new JMenuBar();
        JMenu file = new JMenu("File");
        JMenuItem run  = new JMenuItem("Run");
        
        JMenuItem exit   = new JMenuItem("Exit");
              
        this.setJMenuBar(barraMenus);
        barraMenus.add(file);
        barraMenus.add(settings);
        file.add(run);
        file.add(exit);
        settings.add(motherShip);
        settings.add(obstacle);
        settings.add(sample);
            
        robot1 = new ImageIcon("Reactivos\\imagenes\\wall-e.png");
        robot1 = new ImageIcon(robot1.getImage().getScaledInstance(50,50,  java.awt.Image.SCALE_SMOOTH));
        
        robot2 = new ImageIcon("Reactivos\\imagenes\\eva.png");
        robot2 = new ImageIcon(robot2.getImage().getScaledInstance(50,50,  java.awt.Image.SCALE_SMOOTH));
        
        obstacleIcon = new ImageIcon("Reactivos\\imagenes\\brick.png");
        obstacleIcon = new ImageIcon(obstacleIcon.getImage().getScaledInstance(50,50,  java.awt.Image.SCALE_SMOOTH));
        
        sampleIcon = new ImageIcon("Reactivos\\imagenes\\sample.png");
        sampleIcon = new ImageIcon(sampleIcon.getImage().getScaledInstance(50,50,  java.awt.Image.SCALE_SMOOTH));
        
        motherIcon = new ImageIcon("Reactivos\\imagenes\\mothership.png");
        motherIcon = new ImageIcon(motherIcon.getImage().getScaledInstance(50,50,  java.awt.Image.SCALE_SMOOTH));
        
        this.setLayout(null);
        formaPlano();  
        
        exit.addActionListener(evt -> gestionaSalir(evt));
        run.addActionListener(evt -> gestionaRun(evt));
        obstacle.addItemListener(evt -> gestionaObstacle(evt));
        sample.addItemListener(evt -> gestionaSample(evt));
        motherShip.addItemListener(evt -> gestionaMotherShip(evt));

              
            
        class MyWindowAdapter extends WindowAdapter
        {
            @Override
            public void windowClosing(WindowEvent eventObject)
            {
		goodBye();
            }
        }
        addWindowListener(new MyWindowAdapter());
        
        // Crea 2 agentes
        wallE = new Agente("Wall-E",robot1, matrix, tablero); 
        eva = new Agente("Eva",robot2, matrix, tablero); 
        
    }
        
    private void gestionaSalir(ActionEvent eventObject)
    {
        goodBye();
    }
        
    private void goodBye()
    {
        int respuesta = JOptionPane.showConfirmDialog(rootPane, "Desea salir?","Aviso",JOptionPane.YES_NO_OPTION);
        if(respuesta==JOptionPane.YES_OPTION) System.exit(0);
    }
  
    private void formaPlano()
    {
        tablero = new JLabel[dim][dim];
        matrix = new int[dim][dim];
        
        int i, j;
        
        for(i=0;i<dim;i++)
            for(j=0;j<dim;j++)
            {
                matrix[i][j]=0;
                tablero[i][j]=new JLabel();
                tablero[i][j].setBounds(j*50+10,i*50+10,50,50);
                tablero[i][j].setBorder(BorderFactory.createDashedBorder(Color.white));
                tablero[i][j].setOpaque(false);
                this.add(tablero[i][j]);
                
                tablero[i][j].addMouseListener(new MouseAdapter() // Este listener nos ayuda a agregar poner objetos en la rejilla
                    {
                        @Override
                        public void mousePressed(MouseEvent e) 
                        {
                            insertaObjeto(e);
                        }   
                
                        @Override
                        public void mouseReleased(MouseEvent e) 
                        {
                                insertaObjeto(e);
                        }   
                
                    });
                                
            }
    }
        
    private void gestionaObstacle(ItemEvent eventObject)
    {
        JRadioButtonMenuItem opt = (JRadioButtonMenuItem) eventObject.getSource();
        if(opt.isSelected())
            actualIcon = obstacleIcon;
        else actualIcon = null;        
    }
    
    private void gestionaSample(ItemEvent eventObject)
    {
        JRadioButtonMenuItem opt = (JRadioButtonMenuItem) eventObject.getSource();
        if(opt.isSelected())
           actualIcon = sampleIcon;
        else actualIcon = null;   
    }
    
    private void gestionaMotherShip(ItemEvent eventObject)
    {
        JRadioButtonMenuItem opt = (JRadioButtonMenuItem) eventObject.getSource();
        if(opt.isSelected())
           actualIcon = motherIcon;
        else actualIcon = null;   
    }
    private void gestionaRun(ActionEvent eventObject)
    {
        if(!wallE.isAlive()) wallE.start();
        if(!eva.isAlive()) eva.start();
        settings.setEnabled(false);
    }
       
    public void insertaObjeto(MouseEvent e)
    {
        JLabel casilla = (JLabel) e.getSource();
        int i = (casilla.getY() - 10) / 50;  // Fila en la matriz
        int j = (casilla.getX() - 10) / 50;  // Columna en la matriz
        if(actualIcon!=null){
            casilla.setIcon(actualIcon);
            if (actualIcon == obstacleIcon){
                matrix[i][j] = 1;
            } else if (actualIcon == sampleIcon){
                matrix[i][j] = 2;
            } else if (actualIcon == motherIcon){
                matrix[i][j] = 3;
            }
        }
    }
    
}

File: practica1/Escenario.py
import pygame
import sys
import random
from pygame.locals import *
from Agente import Agente  # Importar la clase Agente desde el archivo Agente.py

class Escenario:
    def __init__(self):
        pygame.init()
        self.dim = 15
        self.cell_size = 50
        self.width = self.dim * self.cell_size + 35
        self.height = self.dim * self.cell_size + 85
        self.screen = pygame.display.set_mode((self.width, self.height))
        pygame.display.set_caption("Agentes")
        self.clock = pygame.time.Clock()
        self.background = pygame.image.load("imagenes/surface.jpg")
        self.background = pygame.transform.scale(self.background, (self.width, self.height))
        self.tablero = [[None for _ in range(self.dim)] for _ in range(self.dim)]
        self.matrix = [[0 for _ in range(self.dim)] for _ in range(self.dim)]
        self.robot1 = pygame.image.load("imagenes/wall-e.png")
        self.robot1 = pygame.transform.scale(self.robot1, (self.cell_size, self.cell_size))
        self.robot2 = pygame.image.load("imagenes/eva.png")
        self.robot2 = pygame.transform.scale(self.robot2, (self.cell_size, self.cell_size))
        self.obstacleIcon = pygame.image.load("imagenes/brick.png")
        self.obstacleIcon = pygame.transform.scale(self.obstacleIcon, (self.cell_size, self.cell_size))
        self.sampleIcon = pygame.image.load("imagenes/sample.png")
        self.sampleIcon = pygame.transform.scale(self.sampleIcon, (self.cell_size, self.cell_size))
        self.motherIcon = pygame.image.load("imagenes/mothership.png")
        self.motherIcon = pygame.transform.scale(self.motherIcon, (self.cell_size, self.cell_size))
        self.actualIcon = None
        self.wallE = Agente("Wall-E", self.robot1, self.matrix, self.tablero)
        self.eva = Agente("Eva", self.robot2, self.matrix, self.tablero)
        self.run()

    def run(self):
        while True:
            self.screen.blit(self.background, (0, 0))
            for event in pygame.event.get():
                if event.type == QUIT:
                    self.goodBye()
                elif event.type == MOUSEBUTTONDOWN:
                    self.insertaObjeto(event.pos)
            self.draw_grid()
            self.wallE.update()
            self.eva.update()
            pygame.display.flip()
            self.clock.tick(30)

    def draw_grid(self):
        for i in range(self.dim):
            for j in range(self.dim):
                rect = pygame.Rect(j * self.cell_size + 10, i * self.cell_size + 10, self.cell_size, self.cell_size)
                pygame.draw.rect(self.screen, (255, 255, 255), rect, 1)
                if self.tablero[i][j]:
                    self.screen.blit(self.tablero[i][j], rect.topleft)

    def insertaObjeto(self, pos):
        x, y = pos
        i = (y - 10) // self.cell_size
        j = (x - 10) // self.cell_size
        if 0 <= i < self.dim and 0 <= j < self.dim and self.actualIcon:
            self.tablero[i][j] = self.actualIcon

    def goodBye(self):
        pygame.quit()
        sys.exit()

if __name__ == "__main__":
    escenario = Escenario()